apps/worker/src/app/workflow/usecases/send-message/send-message-push.usecase.ts

```diff
@@ -113,81 +120,103 @@ export class SendMessagePush extends SendMessageBase {
         Object.values(PushProviderIdEnum).includes(chan.providerId as PushProviderIdEnum)
       ) || [];
 
-    const messagePayload = Object.assign({}, command.payload);
-    delete messagePayload.attachments;
-
     if (!pushChannels.length) {
-      await this.sendErrors(command);
+      await this.sendNoActiveChannelError(command.job);
+      await this.sendNotificationError(command.job);
 
       return;
     }
 
-    for (const channel of pushChannels) {
-      if (!channel.credentials?.deviceTokens) {
-        await this.createExecutionDetails.execute(
-          CreateExecutionDetailsCommand.create({
-            ...CreateExecutionDetailsCommand.getDetailsFromJob(command.job),
-            detail: DetailEnum.PUSH_MISSING_DEVICE_TOKENS,
-            source: ExecutionDetailsSourceEnum.INTERNAL,
-            status: ExecutionDetailsStatusEnum.FAILED,
-            isTest: false,
-            isRetry: false,
-          })
-        );
-
-        continue;
-      }
+    const messagePayload = Object.assign({}, command.payload);
+    delete messagePayload.attachments;
 
-      const integration = await this.getIntegration({
-        id: channel._integrationId,
-        organizationId: command.organizationId,
-        environmentId: command.environmentId,
-        channelType: ChannelTypeEnum.PUSH,
-        providerId: channel.providerId,
-        userId: command.userId,
-        filterData: {
-          tenant: command.job.tenant,
-        },
-      });
+    let integrationsWithErrors = 0;
+    for (const channel of pushChannels) {
+      const { deviceTokens } = channel.credentials || {};
 
-      if (!integration) {
-        await this.createExecutionDetails.execute(
-          CreateExecutionDetailsCommand.create({
-            ...CreateExecutionDetailsCommand.getDetailsFromJob(command.job),
-            detail: DetailEnum.SUBSCRIBER_NO_ACTIVE_INTEGRATION,
-            source: ExecutionDetailsSourceEnum.INTERNAL,
-            status: ExecutionDetailsStatusEnum.FAILED,
-            isTest: false,
-            isRetry: false,
-          })
-        );
+      const [isChannelMissingDeviceTokens, integration] = await Promise.all([
+        this.isChannelMissingDeviceTokens(channel, command),
```

Comment:
Unfortunately I need the channel (interface IChannelSettings) to provide the right context to the execution details stored in `sendPushMissingDeviceTokensError`. I can't use the Job providerId because for more than one active push provider it is only storing one of them.
That's another bug we would need to review in a different task that might come from lack of knowledge in the team that for Push multiple providers could be active. 😕

Reply:
missed that one 🤦‍♂️
good point!!